rfcs/2021-10-29-8621-framing-and-codecs-sinks.md

```diff
@@ -0,0 +1,128 @@
+
+## Scope
+
+### In scope
+
+- Common support structures for encoding/framing
+- Set of commonly used encoding (text, JSON, ...) and framing (newline-delimited, octet-counting, ...) methods
+- Strategy for integrating encoding/framing with sinks
+
+### Out of scope
+
+- Integration with schema system (automatic transformation of event fields depending)
+
+## Pain
+
+Internally, we don't share implementations for encoding strategies that are very commonly used (e.g. JSON, newline-delimited). This leads to duplicated effort and missing guidance for handling encoding in sinks since no canonical implementation exists.
+
+Externally, users are restricted by the available custom set of encodings implemented by the sink. There does not exist any mental model for consistent encoding/framing options for event data that applies to all sinks.
+
+## Proposal
+
+### User Experience
+
+Users should be able to set `encoding` and `framing` options on sinks, analogously to the `decoding` and `framing` options on sources. These options uniformly control how the event _payload_ is encoded. This distinction is important, as encoding for the sink specific _protocol_ and the event _payload_ are separate concerns. The payload should still be encoded according to the sink's protocol, and the sink should provide additional options if there are multiple protocols to choose from, e.g. under a `protocol` key. If there need to be any encoding options for the payload at all can be decided on a per-sink basis.
+
+The fields containing event transformations in sinks on the current `encoding` options (`only_fields`, `except_fields`, `timestamp_format`) should also work with the new encoding options without breaking backwards compatibility. In the future, they may be moved to a dedicated key, e.g. `transform` or replaced by a mechanism provided by the upcoming schema support.
+
+### Implementation
+
+The following common config structures will be provided, analogously to the decoding/framing configs:
+
+```rust
+pub trait SerializerConfig: Debug + DynClone + Send + Sync {
+    /// Builds a serializer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedSerializer>;
+}
+```
+
+```rust
+pub trait FramingConfig: Debug + DynClone + Send + Sync {
+    /// Builds a framer from this configuration.
+    ///
+    /// Fails if the configuration is invalid.
+    fn build(&self) -> crate::Result<BoxedFramer>;
+}
+```
+
+These can be built and combined to form an `Encoder`:
+
+```rust
+/// An encoder that can encode structured events to byte messages.
+pub struct Encoder {
+    serializer: BoxedSerializer,
+    framer: BoxedFramer,
+}
+```
+
+`Encoder` implements `tokio_util::codec::Encoder<Event>`. Internally, events first go through the `Serializer` which implements `tokio_util::codec::Encoder<Event>` and are then handed over to the `Framer` which implements `tokio_util::codec::Encoder<Bytes>`.
```

Comment:
Do you imagine this will work in the same way for sinks that don't use one of the pieces (e.g. framing)?

It seems like sinks that entirely do their own serialization and framing (e.g. datadog_logs) would have an easy enough time opting out entirely, but would something like kafka have to pay some performance/complexity overhead of dealing with the possibility of a framer even though that's built into the protocol?

Reply:
Hm, I see your point. The Kafka sink would still be able to add a `SerializerConfig` only, to build the `Serializer` without wrapping it in an `Encoder`.

Otherwise could be worth to do some micro benchmarks for how much of an impact going through one no-op framing operation has. Might not be significant, as the cost only occurs once per frame.

Reply:
Clarified in d7c9cbfe47a40dcc035fab4aea62037c3914647a.